crates/matrix-sdk-common/src/deserialized_responses.rs

```diff
@@ -432,103 +396,195 @@ impl From<DecryptedRoomEvent> for SyncTimelineEvent {
+            // this way, we simply cause the `room_id` field in the json to be
+            // ignored by a subsequent deserialization.
+            inner: TimelineEventInner::PlainText { event: event.cast() },
             push_actions: None,
-            unsigned_encryption_info: None,
         }
     }
 
     /// Returns a reference to the (potentially decrypted) Matrix event inside
     /// this `TimelineEvent`.
     pub fn event(&self) -> &Raw<AnySyncTimelineEvent> {
-        // TODO: make `inner_event` an AnySyncTimelineEvent instead.
-        self.inner_event.cast_ref()
+        self.inner.event()
     }
 
     /// If the event was a decrypted event that was successfully decrypted, get
     /// its encryption info. Otherwise, `None`.
     pub fn encryption_info(&self) -> Option<&EncryptionInfo> {
-        self.inner_encryption_info.as_ref()
+        self.inner.encryption_info()
     }
 
     /// Takes ownership of this `TimelineEvent`, returning the (potentially
     /// decrypted) Matrix event within.
     pub fn into_event(self) -> Raw<AnySyncTimelineEvent> {
-        // TODO: make `inner_event` an AnySyncTimelineEvent instead.
-        self.inner_event.cast()
+        self.inner.into_event()
     }
 }
 
 impl From<DecryptedRoomEvent> for TimelineEvent {
     fn from(decrypted: DecryptedRoomEvent) -> Self {
-        Self {
-            // Casting from the more specific `AnyMessageLikeEvent` (i.e. an event without a
-            // `state_key`) to a more generic `AnyTimelineEvent` (i.e. one that may contain
-            // a `state_key`) is safe.
-            inner_event: decrypted.event.cast(),
-            inner_encryption_info: Some(decrypted.encryption_info),
-            push_actions: None,
-            unsigned_encryption_info: decrypted.unsigned_encryption_info,
-        }
+        Self { inner: TimelineEventInner::Decrypted(decrypted), push_actions: None }
     }
 }
 
 #[cfg(not(tarpaulin_include))]
 impl fmt::Debug for TimelineEvent {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        let TimelineEvent {
-            inner_event,
-            inner_encryption_info,
-            push_actions,
-            unsigned_encryption_info,
-        } = self;
+        let TimelineEvent { inner, push_actions } = self;
         let mut s = f.debug_struct("TimelineEvent");
-        s.field("event", &DebugRawEvent(inner_event));
-        s.maybe_field("encryption_info", inner_encryption_info);
+        s.field("inner", &inner);
         if let Some(push_actions) = &push_actions {
             if !push_actions.is_empty() {
                 s.field("push_actions", push_actions);
             }
         }
-        s.maybe_field("unsigned_encryption_info", unsigned_encryption_info);
         s.finish()
     }
 }
 
+/// The event within a [`TimelineEvent`] or [`SyncTimelineEvent`], together with
+/// encryption data.
+#[derive(Clone, Serialize, Deserialize)]
+pub enum TimelineEventInner {
```

Comment:
While I'm not a super fan of the `Kind` name scheme, I do prefer it over `Inner`, and think that an enum called `TimelineEventKind` would be better than `Inner` here (especially in a public API). Other ideas that make it clearer that it's about the kind of encryption may work nicely too.

Reply:
35e1bf46e renames to TimelineEventKind